### PR TITLE
[Merged by Bors] - Add wire pkg with ATX type which is serialized for p2p and signing

### PR DESCRIPTION
**Pull request by poszu**
## Motivation
Part of #5774

Split ActivationTx into:
- a type used for encoding and signing, serialized into blobs and sent "over the wire" (`package wire`),
- a type meant for use in the rest of the code (`package types`) - this is not serialized anywhere and is free to change at any time.

## Description

Added a "wire-only" type for the existing V1 ATX. The type in `package types` is no longer used for serialization and signing and can be modified (i.e. for ATX merging) without breakage.

The `types.ActivationTx` will be refactored and simplified in follow-up PRs.

## Test Plan

Existing tests pass.

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: hash/hash.go [hunk 1]
```diff
@@ -21,3 +21,12 @@ func Sum(chunks ...[]byte) (rst [32]byte) {
 	hh.Sum(rst[:0])
 	return rst
 }
+
+func Sum20(chunks ...[]byte) (rst [20]byte) {
+	hh := New()
```
Comment (ivan4th) [inline hash/hash.go]: It would be best to use `sync.Pool` for the hasher here, as if this function is called relatively often, that will incur extra GC pressure b/c the the structure allocated by `New` here is not so small

File: common/types/malfeasance.go [hunk 2]
```diff
@@ -355,17 +350,19 @@ func MalfeasanceInfo(smesher NodeID, mp *MalfeasanceProof) string {
...
+
+		// p.Atx.Initialize()
+		// b.WriteString(
+		// 	fmt.Sprintf(
+		// 		"cause: smesher published ATX %s with invalid post index %d in epoch %d\n",
+		// 		p.Atx.ID().ShortString(),
+		// 		p.InvalidIdx,
+		// 		p.Atx.PublishEpoch,
+		// 	))
+		// }
```
Comment (ivan4th) [inline common/types/malfeasance.go]: Same as above: need to explain the problems encountered here. Also, is the situation with `InvalidPostIndex` just ignored here?

File: common/types/wire/activation_v1.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+func (atx *ActivationTxV1) SignedBytes() []byte {
+	data := codec.MustEncode(&ATXMetadataV1{
+		PublishEpoch: atx.PublishEpoch,
+		MsgHash:      atx.HashInnerBytes(),
+	})
+	return data
+}
+
+func (atx *ActivationTxV1) HashInnerBytes() (result primitive.Hash32) {
+	h := hash.New()
```
Comment (ivan4th) [inline common/types/wire/activation_v1.go]: Using `sync.Pool` for hashers may improve performance by reducing GC pressure

**Review by ivan4th — CHANGES_REQUESTED** [review]

File: activation/activation_multi_test.go [hunk 4]
```diff
@@ -440,9 +439,9 @@ func Test_Builder_Multi_HappyPath(t *testing.T) {
 			func(ctx context.Context, _ string, got []byte) error {
 				atxMtx.Lock()
 				defer atxMtx.Unlock()
-				var gotAtx types.ActivationTx
-				require.NoError(t, codec.Decode(got, &gotAtx))
-				atxs[gotAtx.SmesherID] = gotAtx
+				atx, err := types.AcivationTxFromBytes(got)
```
Comment (fasmat) [inline activation/activation_multi_test.go]: NIT: typo

```suggestion
				atx, err := types.ActivationTxFromBytes(got)
```

File: activation/activation.go [hunk 5]
```diff
@@ -685,7 +685,8 @@ func (b *Builder) createAtx(
 }
 
 func (b *Builder) broadcast(ctx context.Context, atx *types.ActivationTx) (int, error) {
-	buf, err := codec.Encode(atx)
+	// TODO: in future, encode the right ATX version depending on the epoch.
+	buf, err := codec.Encode(atx.ToWireV1())
```
Comment (fasmat) [inline activation/activation.go]: I feel like the `wire.ActivationTxV1` type should be in the `activation` package. Only the `activation.Builder` and `activation.Handler` need it to encode/decode an ATX to be broadcasted (for gossip we just gossip the bytes without even knowing how the encode an ATX).

So instead of `ToWireV1` being a method of `types.ActivationTx` it should be a function in the `activation` package that takes an `types.ActivationTx` as parameter and returns the ATX v1 wire type (which can be a private type of the `activation` package).

File: activation/activation.go [hunk 6]
```diff
@@ -739,7 +740,8 @@ func (b *Builder) Regossip(ctx context.Context, nodeID types.NodeID) error {
 
 // SignAndFinalizeAtx signs the atx with specified signer and calculates the ID of the ATX.
 func SignAndFinalizeAtx(signer *signing.EdSigner, atx *types.ActivationTx) error {
-	atx.Signature = signer.Sign(signing.ATX, atx.SignedBytes())
+	// FIXME - there is no need to sign types.ActivationTX (only wire.ActivationTxVx)
```
Comment (fasmat) [inline activation/activation.go]: Yes this is a side effect of having the current activation TX type both be the wire type and structure used everywhere else - the `SignAndFinalizeAtx` function shouldn't even be needed outside the `activation.Builder` service, even in tests.

File: activation/handler.go [hunk 7]
```diff
@@ -562,10 +563,14 @@ func (h *Handler) handleAtx(
...
 	receivedTime := time.Now()
-	var atx types.ActivationTx
-	if err := codec.Decode(msg, &atx); err != nil {
+
+	var atxOnWire wire.ActivationTxV1
+	if err := codec.Decode(msg, &atxOnWire); err != nil {
 		return nil, fmt.Errorf("%w: %w", errMalformedData, err)
 	}
+
+	atx := types.ActivationTxFromWireV1(&atxOnWire)
```
Comment (fasmat) [inline activation/handler.go]: This should be a method of the handler not a function in the `types` package. No other part of the code should even be aware of the `wire.ActivationTxV1` type.

File: common/types/activation.go [hunk 8]
```diff
@@ -525,3 +399,153 @@ type EpochActiveSet struct {
...
+	var wireAtx wire.ActivationTxV1
+	err := codec.Decode(data, &wireAtx)
+	if err != nil {
+		return nil, fmt.Errorf("decoding ATX: %w", err)
+	}
+
+	return ActivationTxFromWireV1(&wireAtx), nil
+}
+
+func ActivationTxFromWireV1(atx *wire.ActivationTxV1) *ActivationTx {
```
Comment (fasmat) [inline common/types/activation.go]: As stated in other comments; this should not be part of the `types` package in my opinion. No part of the node needs to be aware of how ATXs are encoded or how to decode them; only the `activation` package needs to know this.

File: common/types/activation.go [hunk 9]
```diff
@@ -525,3 +399,153 @@ type EpochActiveSet struct {
...
+			VRFNonce:          (*wire.VRFPostIndex)(a.VRFNonce),
+		},
+		SmesherID: Hash32(a.SmesherID),
+		Signature: a.Signature,
+	}
+}
+
+// Decode ActivationTx from bytes.
+// In future it should decide which version of ActivationTx to decode based on the publish epoch.
+func AcivationTxFromBytes(data []byte) (*ActivationTx, error) {
```
Comment (fasmat) [inline common/types/activation.go]: Same - this should be in the `activation` package and not here.

File: common/types/activation.go [hunk 10]
```diff
@@ -525,3 +399,153 @@ type EpochActiveSet struct {
...
+			NIPost:   NIPostFromWireV1(atx.NIPost),
+			NodeID:   (*NodeID)(atx.NodeID),
+			VRFNonce: (*VRFPostIndex)(atx.VRFNonce),
+		},
+		SmesherID: NodeID(atx.SmesherID),
+		Signature: atx.Signature,
+	}
+}
+
+func NIPostChallengeFromWireV1(ch wire.NIPostChallengeV1) *NIPostChallenge {
```
Comment (fasmat) [inline common/types/activation.go]: This isn't needed any more - the `sql/localsql/nipost_state.go` file can now be deleted as well as the in code migrations for `localsql` - at least 0001 - 0003.

We just have to add information to the release notes that people upgrading to the next version (imho this PR belongs into v1.5) need to first have upgraded to 1.4.x and not skip that version.

File: common/types/activation.go [hunk 11]
```diff
@@ -525,3 +399,153 @@ type EpochActiveSet struct {
...
+		PublishEpoch:   EpochID(ch.PublishEpoch),
+		Sequence:       ch.Sequence,
+		PrevATXID:      ATXID(ch.PrevATXID),
+		PositioningATX: ATXID(ch.PositioningATX),
+		CommitmentATX:  (*ATXID)(ch.CommitmentATX),
+		InitialPost:    PostFromWireV1(ch.InitialPost),
+	}
+}
+
+func NIPostFromWireV1(nipost *wire.NIPostV1) *NIPost {
```
Comment (fasmat) [inline common/types/activation.go]: This can be inlined

File: common/types/activation.go [hunk 12]
```diff
@@ -525,3 +399,153 @@ type EpochActiveSet struct {
...
+		Membership: *MerkleProofFromWireV1(nipost.Membership),
+		Post:       PostFromWireV1(nipost.Post),
+		PostMetadata: &PostMetadata{
+			Challenge:     nipost.PostMetadata.Challenge,
+			LabelsPerUnit: nipost.PostMetadata.LabelsPerUnit,
+		},
+	}
+}
+
+func PostFromWireV1(post *wire.PostV1) *Post {
```
Comment (fasmat) [inline common/types/activation.go]: This can also be removed if we remove the go migrations

File: common/types/activation.go [hunk 13]
```diff
@@ -525,3 +399,153 @@ type EpochActiveSet struct {
...
+		return nil
+	}
+	return &Post{
+		Nonce:   post.Nonce,
+		Indices: post.Indices,
+		Pow:     post.Pow,
+	}
+}
+
+func MerkleProofFromWireV1(proofV1 wire.MerkleProofV1) *MerkleProof {
```
Comment (fasmat) [inline common/types/activation.go]: Same as for the others - remove go migrations and this function shouldn't be needed any more.

File: common/types/hashes.go [hunk 14]
```diff
@@ -1,110 +1,27 @@
...
-const (
-	// Hash32Length is 32, the expected length of the hash.
-	Hash32Length = 32
-	hash20Length = 20
-	hash12Length = 12
+// Reexports.
+type (
+	Hash20 = primitive.Hash20
+	Hash32 = primitive.Hash32
 )
```
Comment (fasmat) [inline common/types/hashes.go]: Why move the type definitions to a different package and alias them here? Can't they stay here? I assume this is because of the `wire` package? This problem goes away if the types in the `wire` package are moved to the `activation` package (or a sub package of it)

File: common/types/hashes.go [hunk 15]
```diff
@@ -1,110 +1,27 @@
...
-	h32 := hash.Sum(data)
-	copy(h[:], h32[:])
-	return
-}
+var (
+	BytesToHash = primitive.BytesToHash
+	CalcHash20  = primitive.CalcHash20
+	CalcHash32  = primitive.CalcHash32
+	HexToHash32 = primitive.HexToHash32
+)
```
Comment (fasmat) [inline common/types/hashes.go]: Re-exports are a symptomatic fix of having types are in packages where they don't belong 🙂

**Review by fasmat — CHANGES_REQUESTED** [review]

File: common/types/malfeasance.go (hunk 2 above)
Comment (fasmat) [reply common/types/malfeasance.go]: I assume these will be fixed before the PR is merged - @poszu maybe mark the PR as draft in the meantime?

File: common/types/primitive/hashes.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
+package primitive
```
Comment (fasmat) [inline common/types/primitive/hashes.go]: The reason this package exists is because otherwise we would have a cyclic dependency between `types` and `wire`

In my opinion the fix for this is to move the types in `wire` to the `activation` package and allow them to import `types` but `types` must not depend on `activation`. That's why converters from `ActivatonTx` to `wire.V1` cannot be methods on the `ActivationTx` type but have to be functions in the `activation` package (and should also only be used there).

File: common/types/testutil.go [hunk 17]
```diff
@@ -66,7 +66,7 @@ func RandomNodeID() NodeID {
 
 // RandomBallotID generates a random BallotID for testing.
 func RandomBallotID() BallotID {
-	var b [hash20Length]byte // TODO(mafa): BallotIDSize is 32???
+	var b [Hash20Length]byte // TODO(mafa): BallotIDSize is 32???
```
Comment (fasmat) [inline common/types/testutil.go]: This is something that I don't understand. Afaik the BallotID is a Hash20 (actually a Hash32 where 12 bytes are cut off) but it is then encoded with 32 bytes - essentially causing every ballot to have 12 0-bytes in its ID?

The same seems to be the case for Proposals

File: common/types/wire/doc.go [hunk 18]
```diff
@@ -0,0 +1,6 @@
+/*
+Package wire contains types that are encoded for:
+- sending over the network
+- signing
+*/
```
Comment (fasmat) [inline common/types/wire/doc.go]: This isn't easily achievable. The packages containing the `Handlers` and `Builders` should contain their respective wire types in them. ATXs (or their wire types) should be in the `activation` package. Proposal wire types should be in the `proposals` package, etc.

`common` should be reserved for domain objects, i.e. the types used by the services but not used for encoding.

File: common/types/wire/malfeasance.go [hunk 19]
```diff
@@ -0,0 +1,10 @@
+package wire
+
+//go:generate scalegen
+
+type InvalidPostIndexProofV1 struct {
```
Comment (fasmat) [inline common/types/wire/malfeasance.go]: We should consider moving this type to the `malfeasance` package (along other types for malfeasance proofs)

File: genvm/templates/multisig/types.go [hunk 20]
```diff
@@ -4,6 +4,8 @@ import (
 	"bytes"
 	"fmt"
 
+	// required to help scalegen find the types.
+	_ "github.com/spacemeshos/go-spacemesh/common/types/primitive"
```
Comment (fasmat) [inline genvm/templates/multisig/types.go]: This should not be needed - assuming we can inline the primitive package

File: sql/localsql/0001_migration_test.go [hunk 21]
```diff
@@ -16,14 +16,14 @@ import (
 )
 
 func saveNipostChallenge(dir string, ch *types.NIPostChallenge) error {
-	if err := save(filepath.Join(dir, challengeFilename), ch); err != nil {
+	if err := save(filepath.Join(dir, challengeFilename), ch.ToWireV1()); err != nil {
```
Comment (fasmat) [inline sql/localsql/0001_migration_test.go]: I think we can remove the 0001 - 0003 migrations with this PR (assuming this change goes into a v1.5.x version of the node).

File: sql/localsql/nipost/nipost.go [hunk 22]
```diff
@@ -16,7 +17,7 @@ type NIPostState struct {
 }
 
 func AddNIPost(db sql.Executor, nodeID types.NodeID, nipost *NIPostState) error {
-	buf, err := codec.Encode(&nipost.Membership)
+	buf, err := codec.Encode(nipost.Membership.ToWireV1())
```
Comment (fasmat) [inline sql/localsql/nipost/nipost.go]: I think originally I did this because I was lazy and didn't want to map the `Membership` object to columns of the table, which I still think is more effort than what it is worth.

Yet we must not have any wire types in our `sql` functions, so I believe we need a dedicated type here, that is only used to serialize the membership for storing the DB and deserialze it when fetching it again. We cannot have our `sql` package depend on actual wire types.

File: sql/localsql/nipost_state_types.go [hunk 23]
```diff
@@ -20,7 +23,7 @@ type PoetRequest struct {
 type NIPostBuilderState struct {
 	Challenge types.Hash32
 
-	NIPost *types.NIPost
+	NIPost *wire.NIPostV1
```
Comment (fasmat) [inline sql/localsql/nipost_state_types.go]: No wire types in database objects.

File: systest/tests/distributed_post_verification_test.go [hunk 24]
```diff
@@ -280,12 +281,12 @@ func TestPostMalfeasanceProof(t *testing.T) {
...
 		require.Equal(t, *atx.CommitmentATX, *invalidAtx.CommitmentATX)
-		require.Equal(t, atx.NIPostChallenge, invalidAtx.NIPostChallenge)
+		require.Equal(t, atx.NIPostChallenge, invalidAtx.NIPostChallengeV1)
 		require.Equal(t, atx.NIPost.Post.Indices, invalidAtx.NIPost.Post.Indices)
 
 		meta := &shared.ProofMetadata{
-			NodeId:          invalidAtx.NodeID.Bytes(),
-			CommitmentAtxId: invalidAtx.CommitmentATX.Bytes(),
+			NodeId:          invalidAtx.NodeID[:],
+			CommitmentAtxId: invalidAtx.CommitmentATX[:],
```
Comment (fasmat) [inline systest/tests/distributed_post_verification_test.go]: Why the change? I know functionally there is no difference, but I think `.Bytes()` is easier to understand (for someone not familiar with the code).

**Review by fasmat — CHANGES_REQUESTED** [review]

File: hash/hash.go (hunk 1 above)
Comment (poszu) [reply hash/hash.go]: I opened https://github.com/spacemeshos/go-spacemesh/issues/5813 to track this.

File: common/types/malfeasance.go (hunk 2 above)
Comment (poszu) [reply common/types/malfeasance.go]: Thanks, I forgot about it.

File: common/types/wire/activation_v1.go (hunk 3 above)
Comment (poszu) [reply common/types/wire/activation_v1.go]: IMHO, this would be a premature optimization for this case. ATX is hashed once every epoch, there's no point in optimizing it.

Comment (fasmat) [reply common/types/wire/activation_v1.go]: I believe @ivan4th is referring to the fact that every time an ATX is signed/verified it needs to be hashed, so if many ATXs are verified in a short time period many instances of `hash.Hash` are created which slows down the node and could be improved by re-using existing instances with a `sync.Pool`. I remember when we were still creating PoST proofs in Go we also used a `sync.Pool` for that reason (also for the `hash.Hash` type if I remember correctly).

I agree, that this change doesn't need to be included in this PR but would be a good improvement to do separatly 🙂

File: activation/activation.go (hunk 5 above)
Comment (poszu) [reply activation/activation.go]: > Only the `activation.Builder` and `activation.Handler` need it to encode/decode an ATX to be broadcasted (for gossip we just gossip the bytes without even knowing how the encode an ATX).

Ideally yes, but in reality these types are also (de)en-coded:
- when loading ATX from DB (https://github.com/spacemeshos/go-spacemesh/blob/817fd0e60a1f4a73ed5cbf73828d8aea02376d0f/sql/atxs/atxs.go#L25-L39)
- when migrating from files to localdb:
  * https://github.com/spacemeshos/go-spacemesh/blob/8cbbb1af97ae38202c2d1a7847fb501cf336ed75/sql/localsql/nipost_state.go#L117-L123
  * https://github.com/spacemeshos/go-spacemesh/blob/da627d288663608f1b4c527459fc913d86a35545/sql/localsql/0003_migration.go#L196-L199

Comment (fasmat) [reply activation/activation.go]: I know that's why I was proposing here to go for smaller steps: https://github.com/spacemeshos/go-spacemesh/pull/5789#discussion_r1555362673

The first step should be the introduction of the wire type and use it only in `activation.Handler` and `activation.Builder` and not changing any `sql` functions yet. The next step would be to update `sql` functions to not use go-scale encoding at all (or in the case of the nipost state have their own type just for that).

File: activation/handler.go (hunk 7 above)
Comment (poszu) [reply activation/handler.go]: I also need it in the SQL code to deserialize the blob. Hopefully, we will remove decoding the blob soon and then it can be moved to the ATX handler code.

File: common/types/activation.go (hunk 8 above)
Comment (poszu) [reply common/types/activation.go]: I also need it in the SQL code to deserialize the blob. Hopefully, we will remove decoding the blob soon and then it can be moved to the ATX handler code.

Comment (fasmat) [reply common/types/activation.go]: I want to keep the change much smaller, a lot of the code here is only needed to be removed again in the next PR.

File: common/types/wire/activation_v1.go (hunk 3 above)
Comment (poszu) [reply common/types/wire/activation_v1.go]: True - I forgot that the ATX signature is over (Publish epoch; ATX ID), not the ATX blob.

File: common/types/activation.go (hunk 11 above)
Comment (poszu) [reply common/types/activation.go]: I want it separated as there is an additional nil check. I will un-export it though

File: common/types/activation.go (hunk 13 above)
Comment (poszu) [reply common/types/activation.go]: It's needed as localDB keeps a serialized form of the merkle proof.

File: common/types/hashes.go (hunk 14 above)
Comment (poszu) [reply common/types/hashes.go]: reverted

File: common/types/wire/activation_v1.go (hunk 3 above)
Comment (fasmat) [reply common/types/wire/activation_v1.go]: For ATX v2 instead of having `Publish epoch || ATX ID` as input for the signature we should probably have all fields of the ATX in a tree and use that as payload that is signed; similar to how the `MerkleProof` for PoET works - this way we only need to proof that the specific field(s) we are interested in in a malfeasance proof had the value that was relevant for the proof and everything else is just a hash.

File: systest/tests/distributed_post_verification_test.go (hunk 24 above)
Comment (poszu) [reply systest/tests/distributed_post_verification_test.go]: reverted

File: common/types/activation.go (hunk 13 above)
Comment (poszu) [reply common/types/activation.go]: removed

File: common/types/activation.go (hunk 12 above)
Comment (poszu) [reply common/types/activation.go]: I want to keep it as it simplifies decoding InitialPost (which can be nil) and normal Post.

File: common/types/activation.go (hunk 8 above)
Comment (poszu) [reply common/types/activation.go]: Some code (especially decoding the wire type to types.ActivationTx) needs to exist now and will be removed when types.ActivationTx is simplified to not contain things like Post.

File: activation/activation.go [hunk 25]
```diff
@@ -638,8 +639,10 @@ func (b *Builder) createAtx(
 	challenge *types.NIPostChallenge,
 ) (*types.ActivationTx, error) {
 	pubEpoch := challenge.PublishEpoch
+	// TODO: in future, encode the right NiPoST challenge version depending on the pubEpoch.
+	challengeHash := wire.NIPostChallengeToWireV1(challenge).Hash()
```
Comment (fasmat) [inline activation/activation.go:643]: There might be a transitional period where the builder needs to now both versions but after that switch passed we can remove the logic to build a V1 ATX. 🙂

File: activation/activation.go [hunk 26]
```diff
@@ -758,9 +762,12 @@ func (b *Builder) Regossip(ctx context.Context, nodeID types.NodeID) error {
...
 func SignAndFinalizeAtx(signer *signing.EdSigner, atx *types.ActivationTx) error {
-	atx.Signature = signer.Sign(signing.ATX, atx.SignedBytes())
+	// FIXME - there is no need to sign types.ActivationTX (only ActivationTxVx)
+	wireAtx := wire.ActivationTxToWireV1(atx)
+	atx.Signature = signer.Sign(signing.ATX, wireAtx.SignedBytes())
 	atx.SmesherID = signer.NodeID()
-	return atx.Initialize()
+	atx.SetID(types.ATXID(wireAtx.HashInnerBytes()))
+	return nil
 }
```
Comment (fasmat) [inline activation/activation.go:771]: I would leave this function as it is (maybe mark it deprecated) and change the `Builder` to sign the wire type directly during building. We want to avoid the conversion ATX -> wire.ATX completely; the builder should only know the wire type - the handler also only knows the wire type, verifies it and as the last step after successful verification convert it to the domain type and store it in the DB.

File: activation/handler.go (hunk 7 above)
Comment (fasmat) [reply activation/handler.go]: I think the `ActivationTxFromWireV1` and `ActivationTxFromWireV1` generally should not be needed. We should never convert from Domain to wire type (the builder works on the wire type directly) and the conversion from wire to domain is only needed in the handler in one place - after successful validation and before storing the ATX in the DB.

All tests should either only use the wire type (e.g. Builder and Handler test) or only use the domain type (pretty much everything else that needs ATXs)

File: activation/handler.go [hunk 27]
```diff
@@ -241,7 +242,7 @@ func (h *Handler) SyntacticallyValidateDeps(
 		baseTickHeight = posAtx.TickHeight()
 	}
 
-	expectedChallengeHash := atx.NIPostChallenge.Hash()
+	expectedChallengeHash := wire.NIPostChallengeToWireV1(&atx.NIPostChallenge).Hash()
```
Comment (fasmat) [inline activation/handler.go:245]: Ideally the `SyntacticallyValidate` method accepts a `wire.ActivationTxV1` argument and not a `types.ActivationTx` (as well as all other methods used to handle incoming ATXs). Converting from wire to domain type should only be necessary once - after the ATX was successfully validated and just before it is stored in the DB.

Since the handler will always need to be able to validate V1 and V2 ATXs we will need validation methods for both kind of ATXs. I think the cleanest way to go here is to have the handler receive the raw bytes (from fetch or gossip) decode the first few bytes as pubepoch and then based on those call validation methods of the right (sub-) service that validate the specific ATX type for that public epoch.

Wdyt?

File: activation/wire/wire_v1.go [hunk 28]
```diff
@@ -0,0 +1,240 @@
...
+	"github.com/spacemeshos/go-spacemesh/hash"
+)
+
+//go:generate scalegen
+
+type ActivationTxV1 struct {
+	InnerActivationTxV1
+
+	SmesherID types.NodeID
+	Signature [64]byte
```
Comment (fasmat) [inline activation/wire/wire_v1.go]: I think this could be

```suggestion
	Signature types.EdSignature
```

`types.EdSignature` is a `[64]byte` but giving it a concrete name prevents other `[64]bytes` from being accidentally used as signatures

File: checkpoint/recovery_test.go [hunk 29]
```diff
@@ -255,7 +256,7 @@ func validateAndPreserveData(
 	)
 	mfetch.EXPECT().GetAtxs(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
 	for i, vatx := range deps {
-		encoded, err := codec.Encode(vatx)
+		encoded, err := codec.Encode(wire.ActivationTxToWireV1(vatx.ActivationTx))
```
Comment (fasmat) [inline checkpoint/recovery_test.go:259]: Recovery will not use the domain type but the wire type in my opinion. Meaning that if a checkpoint is created it will only contain the ATX blobs which recovery decodes to the wire type (and then restores by passing it by the activation handler).

Recovery only needs to make sure that the ATXs are persisted in publication order so that the handler can verify them correctly during restore.

File: cmd/merge-nodes/internal/merge_action_test.go [hunk 30]
```diff
@@ -227,13 +228,13 @@ func Test_MergeDBs_Successful_Existing_Node(t *testing.T) {
 	require.NoError(t, err)
 
 	sig1Poet1 := nipost.PoETRegistration{
-		ChallengeHash: sig1Ch.Hash(),
+		ChallengeHash: wire.NIPostChallengeToWireV1(sig1Ch).Hash(),
```
Comment (fasmat) [inline cmd/merge-nodes/internal/merge_action_test.go]: I think you can use a random hash here instead of having to use the wire type in the test (`merge-nodes` shouldn't care about wire types at all).

File: node/node.go [hunk 31]
```diff
@@ -1960,7 +1961,12 @@ func (app *App) verifyDB(ctx context.Context) {
 			}
 
 			// verify atx signature
-			if !app.edVerifier.Verify(signing.ATX, atx.SmesherID, atx.SignedBytes(), atx.Signature) {
+			// TODO: use atx handler to verify signature
```
Comment (fasmat) [inline node/node.go:1966]: This PR or next? 🙂

File: activation/activation.go (hunk 26 above)
Comment (poszu) [reply activation/activation.go:771]: In this PR I wanted minimal changes in Handler and Builder to make it work. I plan to refactor Builder and Handler in the next step in the following PRs.

File: activation/handler.go (hunk 27 above)
Comment (poszu) [reply activation/handler.go:245]: Yeah. I'm currently refactoring the Handler and Validator methods for a follow up PR.

File: node/node.go (hunk 31 above)
Comment (poszu) [reply node/node.go:1966]: TBH, I'd remove this function from `develop` as it's unused and keep it on a branch (maybe in a draft PR).

File: common/types/wire/activation_v1.go (hunk 3 above)
Comment (fasmat) [reply common/types/wire/activation_v1.go]: Created issues for re-use of `hash.Hash` instances here: https://github.com/spacemeshos/go-spacemesh/issues/5850

Updated ATX v2 issue to contain information bout the ATX ID being calculated from a tree: https://github.com/spacemeshos/go-spacemesh/issues/5774

**Review by fasmat — APPROVED** [review]

File: node/node.go (hunk 31 above)
Comment (fasmat) [reply node/node.go:1966]: You can remove it if you want - it can be re-added if needed

File: checkpoint/recovery_test.go (hunk 29 above)
Comment (fasmat) [reply checkpoint/recovery_test.go:259]: This also came up in a discussion on a different PR here: https://github.com/spacemeshos/go-spacemesh/pull/5791#discussion_r1561281266

In one of the next PRs we need to update checkpoint to store the ATXs just as bytes and not wire types.

File: activation/activation.go (hunk 26 above)
Comment (fasmat) [reply activation/activation.go:771]: I added information about dependencies on the wire types to the issue about ATX versioning here: https://github.com/spacemeshos/go-spacemesh/issues/5774

File: activation/handler.go (hunk 27 above)
Comment (fasmat) [reply activation/handler.go:245]: Added some of the information from the discussion to this issue: https://github.com/spacemeshos/go-spacemesh/issues/5774

File: checkpoint/recovery_test.go (hunk 29 above)
Comment (fasmat) [reply checkpoint/recovery_test.go:259]: Also added that information to https://github.com/spacemeshos/go-spacemesh/issues/5774

File: activation/handler.go (hunk 7 above)
Comment (fasmat) [reply activation/handler.go]: Added the info about only `activation` using wire types to https://github.com/spacemeshos/go-spacemesh/issues/5774

**Review by fasmat — APPROVED** [review]

**Comment by poszu** [comment]
bors merge

**Review by ivan4th — APPROVED** [review]

**Comment by poszu** [comment]
bors merge

**Comment by poszu** [comment]
bors merge

**Comment by poszu** [comment]
bors merge
